lib/browser/mqtt.ts

```diff
@@ -159,6 +159,8 @@ export class MqttClientConnection extends BufferedEventEmitter {
      */
     on(event: 'resume', listener: (return_code: number, session_present: boolean) => void): this;
 
+    on(event: 'publish', listener: (topic: string, payload: Buffer) => void): this;
```

Comment:
I looked into it a bit more. The PUBLISH packet actually is used for both purposes in the MQTT spec (client -> server and server -> client).

mqtt.js called this event "message" and paho calls it "on_message" which is probably where I got that idea

Reply:
Do we allow users to pass NULL callbacks when they subscribe? Since this fires for EVERY message, we probably want to allow that. So that users can put all their message-receiving logic here, if they wish.

That's how mqtt.js behaves. the subscribe callback is optional and "message" fires for every received message.

Paho's behavior is that their "on_message" only fires if the message doesn't have a specific callback registered for its topic-filter

Reply:
We could allow null callbacks, nothing really stops us. I'll do a follow up. I'll also swap to on('message'). I was debating it all along, I zigged, shoulda zagged.